### PR TITLE
Feature/basic functionality

### PR DESCRIPTION
**Pull request by ikbendewilliam**

### CODE REVIEW

File: lib/custom_image_crop_widget.dart [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+    canvas.transform(matrix4Image.storage);
+    canvas.drawImage(imageAsUIImage, Offset(-imageAsUIImage.width / 2, -imageAsUIImage.height / 2), paint);
+    canvas.restore();
+    ui.Picture picture = pictureRecorder.endRecording();
+    ui.Image image = await picture.toImage(cropWidth.floor(), cropWidth.floor());
+
+    // Optionally remove magenta from image by evaluating every pixel
+    // See https://github.com/brendan-duncan/image/blob/master/lib/src/transform/copy_crop.dart
+
+    ByteData bytes = await image.toByteData(format: ui.ImageByteFormat.png);
```
Comment (NicolaChimerapps) [inline lib/custom_image_crop_widget.dart:219]: run in isolate

File: lib/custom_image_crop_widget.dart [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+    });
+  }
+
+  bool dataAllowed(CropImageData data) {
+    // For now, this will do. The idea is that we create
+    // a path from the data and check if when we combine
+    // that with the crop path that the resulting path
+    // overlap the hole (crop). So we check if all pixels
+    // from the crop contain pixels from the original image
+    data.scale = data.scale.clamp(0.1, 10.0);
```
Comment (NicolaChimerapps) [inline lib/custom_image_crop_widget.dart:238]: This does nothing given that data+transition will create a new object that is destroyed afterwards

Comment (NicolaChimerapps) [reply lib/custom_image_crop_widget.dart:238]: Also the naming of this function does not convey that it will change the data

File: pubspec.yaml [hunk 3]
```diff
@@ -12,43 +12,10 @@ dependencies:
   flutter:
     sdk: flutter
   gesture_x_detector: ^0.0.5
+  # image: ^2.1.19
```
Comment (NicolaChimerapps) [inline pubspec.yaml:15]: don't keep comments around

**Review by NicolaChimerapps — CHANGES_REQUESTED** [review]
See inline comments

**Comment by vanlooverenkoen** [comment]
Initial Image                        |  Crop Image
:-------------------------:|:-------------------------:
image  |  image

For some reason, the image is not cropped correctly.

**Comment by vanlooverenkoen** [comment]
development issue:
If you start up your application and use a circle as customcropshape everything works fine. 

If you update it to a square and use hot reload to see your new changes this is the result:

image

We should take a look at `void didUpdateWidget(covariant CustomImageCrop oldWidget);` I think the path is not redrawn

File: lib/custom_image_crop_widget.dart [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+    _imageStream?.removeListener(_imageListener);
+    controller?.removeListener(this);
+    super.dispose();
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    width = MediaQuery.of(context).size.width;
+    height = MediaQuery.of(context).size.height;
+    final cropWidth = min(width, height) * widget.cropPercentage;
```
Comment (vanlooverenkoen) [inline lib/custom_image_crop_widget.dart:90]: cropwidth can move after the if statement

File: lib/custom_image_crop_widget.dart [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+  @override
+  void dispose() {
+    _imageStream?.removeListener(_imageListener);
+    controller?.removeListener(this);
+    super.dispose();
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    width = MediaQuery.of(context).size.width;
```
Comment (vanlooverenkoen) [inline lib/custom_image_crop_widget.dart:88]:
```
    return LayoutBuilder(builder: (context, constraints) {
      width = constraints.maxWidth;
      height = constraints.maxHeight;
      ...  
   }  
```

**Comment by vanlooverenkoen** [comment]
The plugin will always use the full width & height. Only take into account the actual space the widget can take up.
(https://github.com/icapps/flutter-custom-image-crop/pull/1/files#diff-36427ab4a1fdca490aa17c986d3e2f197626291632ab97db63993687657d0afeR88) use a layoutbuilder instead

image

File: lib/dotted_path_painter.dart [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+class DottedCropPathPainter extends CustomPainter {
+  Path path;
+
+  DottedCropPathPainter({this.path});
+
+  static Widget drawPath(Path path) => CustomPaint(painter: DottedCropPathPainter(path: path));
+
+  Paint _paint = Paint()
+    ..color = Colors.white
+    ..strokeWidth = 4.0
```
Comment (vanlooverenkoen) [inline lib/dotted_path_painter.dart:14]: use a constant for this

**Review by vanlooverenkoen — CHANGES_REQUESTED** [review]

File: example/lib/main.dart [hunk 7]
```diff
@@ -29,16 +31,53 @@ class MyHomePage extends StatefulWidget {
 }
 
 class _MyHomePageState extends State<MyHomePage> {
+  final controller = CustomImageCropController();
```
Comment (vanlooverenkoen) [inline example/lib/main.dart:34]: create the controller in initstate & dispose the controller in dispose

**Review by vanlooverenkoen — CHANGES_REQUESTED** [review]

File: lib/custom_image_crop_widget.dart (hunk 1 above)
Comment (NicolaChimerapps) [reply lib/custom_image_crop_widget.dart:219]: unclear if this already runs in an isolate. Could be impossible since it cannot be shared across isolates

Comment (ikbendewilliam) [reply lib/custom_image_crop_widget.dart]: Didn't work, so for now we will not implement this. Workaround would include saving and loading the image...

File: pubspec.yaml [hunk 8]
```diff
@@ -16,39 +16,5 @@ dependencies:
 dev_dependencies:
   flutter_test:
     sdk: flutter
-# For information on the generic Dart part of this file, see the
```
Comment (vanlooverenkoen) [inline pubspec.yaml:19]: author can be removed. This is deprecated
homepage should be the homepage of this repo so you should add the github url here

**Review by vanlooverenkoen — CHANGES_REQUESTED** [review]

File: lib/src/dotted_path_painter.dart [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+  static const dashWidth = 10.0;
+  static const dashSpace = 5.0;
+  static const strokeWidth = 4.0;
+  final Path _path;
+
+  DottedCropPathPainter(this._path);
+
+  static Widget drawPath(Path path) => CustomPaint(painter: DottedCropPathPainter(path));
+
+  final _paint = Paint()
```
Comment (vanlooverenkoen) [inline lib/src/dotted_path_painter.dart:15]: fields should be defined before the constructor

File: lib/src/solid_path_painter.dart [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+
+class SolidCropPathPainter extends CustomPainter {
+  static const strokeWidth = 4.0;
+  final Path _path;
+
+  SolidCropPathPainter(this._path);
+
+  static Widget drawPath(Path path) => CustomPaint(painter: SolidCropPathPainter(path));
+
+  final _paint = Paint()
```
Comment (vanlooverenkoen) [inline lib/src/solid_path_painter.dart:13]: Fields should be defined before the constructor

**Review by vanlooverenkoen — CHANGES_REQUESTED** [review]

File: lib/src/custom_image_crop_widget.dart [hunk 11]
```diff
@@ -0,0 +1,249 @@
...
+import 'package:custom_image_crop/src/dotted_path_painter.dart';
+import 'package:custom_image_crop/src/inverted_clipper.dart';
+import 'package:custom_image_crop/src/model.dart';
+
+class CustomImageCrop extends StatefulWidget {
+  final ImageProvider image;
+  final CustomImageCropController cropController;
+  final Color backgroundColor;
+  final Color overlayColor;
+  final Function(double) path;
```
Comment (vanlooverenkoen) [inline lib/src/custom_image_crop_widget.dart:20]: When using a path I get a strange clipped image

code:
```
            child: CustomImageCrop(
              cropController: controller,
              path: (cropWidth) {
                return Path()
                  ..moveTo(50, 50)
                  ..lineTo(100, 50)
                  ..lineTo(100, 100)
                  ..lineTo(50, 100)
                  ..close();
              },
              image: const AssetImage('assets/test.png'), // Any Imageprovider will work, try with a NetworkImage for example...
            ),
```


Image cropper with path             |  Cropped Image
:-------------------------:|:-------------------------:
image  |  image

**Review by vanlooverenkoen — CHANGES_REQUESTED** [review]

File: example/lib/main.dart [hunk 12]
```diff
@@ -14,31 +17,69 @@ class MyApp extends StatelessWidget {
...
   _MyHomePageState createState() => _MyHomePageState();
 }
 
 class _MyHomePageState extends State<MyHomePage> {
+  CustomImageCropController controller;
+
+  @override
+  void initState() {
+    controller = CustomImageCropController();
+    super.initState();
```
Comment (vanlooverenkoen) [inline example/lib/main.dart:40]: start with super.initState()

https://api.flutter.dev/flutter/widgets/State/initState.html
```
If you override this, make sure your method starts with a call to super.initState().
```

**Review by vanlooverenkoen — CHANGES_REQUESTED** [review]

**Comment by vanlooverenkoen** [comment]
Maybe just one more comment about the file structure. Right now the files are split up between the export file & source files. But we should also split up the source files in different directories to keep the same classes together
- models
- painters
- clippers
- widgets

File: lib/src/custom_image_crop_widget.dart (hunk 11 above)
Comment (ikbendewilliam) [reply lib/src/custom_image_crop_widget.dart:20]: Yes I see why. The path is meant to be in regards to cropwidth, width and height. Using const ints like 100 and 50 they're not scaled during the crop to the new width and height, instead the path is being called again and expects to be scaled automatically with the new width and height during the crop 😕 
I could try and scale the path instead or we can add a comment that mentions to always use width and height

Comment (ikbendewilliam) [reply lib/src/custom_image_crop_widget.dart]: removed path for now

**Review by NicolaVerbeeck — APPROVED** [review]

**Pull request merged**